cupyx/_pinned_array.py

```diff
@@ -0,0 +1,157 @@
+        elif x.flags.c_contiguous:
+            order_char = ord('C')
+    return order_char
+
+
+def empty_pinned(shape, dtype=float, order='C'):
+    """Returns an array without initializing the elements.
+
+    This is a convenience function which is just :func:`numpy.empty`,
+    except that the underlying memory is allocated from CuPy's pinned
+    memory pool.
+
+    Args:
+        shape (int or tuple of ints): Dimensionalities of the array.
+        dtype: Data type specifier.
+        order ({'C', 'F'}): Row-major (C-style) or column-major
+            (Fortran-style) order.
+
+    Returns:
+        numpy.ndarray: A new array with elements not initialized.
+
+    .. seealso:: :func:`numpy.empty`
+
+    """
+    shape = _update_shape(None, shape)
+    nbytes = internal.prod(shape) * numpy.dtype(dtype).itemsize
+    mem = cuda.alloc_pinned_memory(nbytes)
+    out = numpy.ndarray(shape, dtype=dtype, buffer=mem, order=order)
+    return out
+
+
+def empty_like_pinned(a, dtype=None, order='K', subok=None, shape=None):
+    """Returns a new array with same shape and dtype of a given array.
+
+    This is a convenience function which is just :func:`numpy.empty_like`,
+    except that the underlying memory is allocated from CuPy's pinned
+    memory pool.
+
+    This function currently does not support ``subok`` option.
+
+    Args:
+        a (numpy.ndarray or cupy.ndarray): Base array.
+        dtype: Data type specifier. The data type of ``a`` is used by default.
+        order ({'C', 'F', 'A', or 'K'}): Overrides the memory layout of the
+            result. ``'C'`` means C-order, ``'F'`` means F-order, ``'A'`` means
+            ``'F'`` if ``a`` is Fortran contiguous, ``'C'`` otherwise.
+            ``'K'`` means match the layout of ``a`` as closely as possible.
+        subok: Not supported yet, must be None.
+        shape (int or tuple of ints): Overrides the shape of the result. If
+            ``order='K'`` and the number of dimensions is unchanged, will try
+            to keep order, otherwise, ``order='C'`` is implied.
+
+    Returns:
+        numpy.ndarray: A new array with same shape and dtype of ``a`` with
+        elements not initialized.
+
+    .. seealso:: :func:`numpy.empty_like`
+
+    """
+    # We're kinda duplicating the code here because order='K' needs special
+    # treatment: strides need to be computed
+    if subok is not None:
+        raise TypeError('subok is not supported yet')
+    if dtype is None:
+        dtype = a.dtype
+    shape = _update_shape(a, shape)
+    order, strides, _ = _new_like_order_and_strides(
+        a, dtype, order, shape, get_memptr=False, get_char=_update_order_char)
+    nbytes = internal.prod(shape) * numpy.dtype(dtype).itemsize
+    mem = cuda.alloc_pinned_memory(nbytes)
+    out = numpy.ndarray(shape, dtype=dtype, buffer=mem,
+                        strides=strides, order=order)
+    return out
+
+
+def zeros_pinned(shape, dtype=float, order='C'):
```

Comment:
I discussed with the dev members and wondering `zeros_*` APIs are really needed or not, because users can simply zero-fill the arrays returned by `empty_*`.

Reply:
I don't have strong opinion here. Let me know if you'd like me to remove zeros*.

I simply followed the NumPy API, which has zeros*, though I do agree when I first learned this I also thought users could just zero it out by themselves.

By the way, I just found PyCUDA also provides both empty* and zeros*:
https://documen.tician.de/pycuda/driver.html?highlight=page%20locked#pagelocked-allocation

Reply:
Thanks for the PyCUDA link, sounds reasonable to provide all of them.